### PR TITLE
Create transaction model and auto create when user pay online

### PR DESCRIPTION
**Pull request by thminhVN**
#### Status

READY

#### What's this PR do?

- [x] Create transaction log model
- [x] Create transaction model
- [x] Auto generate transaction log when user try to pay online
- [x] Auto generate transaction when user pay online successfully
- [x] Send email to user after they pay successfully
- [x] Improve Order.show can find code with "#" before
- [x] Improve email template can show full url belong to host

#### How should this be manually tested?

1. Make new order
2. Update total value in order record
3. Update status to 4
4. Go to /payment?code={code} to pay and check in database

#### What are the relevant Git/YouTrack tickets?

// Put in link to Git Issue or YouTrack ticket

#### Questions:

- Is there a blog post?
- Does the knowledge base need an update?
- Does this add new dependencies which need to be added to?

### CODE REVIEW

**Comment by thminhVN** [comment]
@hieuphq @tieubao plz review it

File: web/controllers/order_controller.ex [hunk 1]
```diff
@@ -291,36 +296,47 @@ defmodule Delivr.OrderController do
...
+
+    if transaction_log do
+      case resp do
+        {:ok, %HTTPoison.Response{status_code: 200, body: body}} ->
+          resp_data = String.replace(body, ~r/(\\n|\\)/, "")
+          {status, result} = JSON.decode(resp_data)
+
+          if (result["status"] == "succeeded") do
+            transaction = generate_transaction_for_order(transaction_log)
+            if transaction do
```
Comment (hieuphq) [inline web/controllers/order_controller.ex]: add empty line before condition statment

File: web/controllers/order_controller.ex [hunk 2]
```diff
@@ -357,4 +377,86 @@ defmodule Delivr.OrderController do
...
+        |> Email.add_to(order.email, order.name)
+        |> Email.put_subject("Đơn hàng #{order.order_code} của bạn đã đặt cọc thành công")
+        |> Email.put_html(
+          Phoenix.View.render_to_string(Delivr.EmailView, "thank_you_paid_online.html", conn: conn, order: order)
+        )
+
+      Mailer.send(email)
+    end
+
+
```
Comment (hieuphq) [inline web/controllers/order_controller.ex]: delivr's email, branch name should config in config.exs like as application config

File: web/models/transaction.ex [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+    field :sender_branch_name, :string
+    field :sender_account_number, :string
+    field :receiver_name, :string
+    field :receiver_gateway_name, :string
+    field :receiver_branch_name, :string
+    field :receiver_account_number, :string
+    belongs_to :transaction_log, Delivr.TransactionLog
+    belongs_to :order, Delivr.Order
+
+    timestamps()
```
Comment (hieuphq) [inline web/models/transaction.ex]: should change to timestamps([{:inserted_at, :created_at}])

File: priv/repo/migrations/20160812145358_create_transaction_log.exs [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+      add :receiver_name, :string
+      add :receiver_gateway_name, :string
+      add :receiver_branch_name, :string
+      add :receiver_account_number, :string
+      add :request_data, :text
+      add :response_data, :text
+      add :transaction_id, :string
+      add :order_id, references(:orders, on_delete: :nothing)
+
+      timestamps()
```
Comment (hieuphq) [inline priv/repo/migrations/20160812145358_create_transaction_log.exs]: should change to timestamps([{:inserted_at, :created_at}])

File: priv/repo/migrations/20160814154456_create_transaction.exs [hunk 5]
```diff
@@ -0,0 +1,27 @@
...
+      add :sender_branch_name, :string
+      add :sender_account_number, :string
+      add :receiver_name, :string
+      add :receiver_gateway_name, :string
+      add :receiver_branch_name, :string
+      add :receiver_account_number, :string
+      add :transaction_log_id, references(:transaction_logs, on_delete: :nothing)
+      add :order_id, references(:orders, on_delete: :nothing)
+
+      timestamps()
```
Comment (hieuphq) [inline priv/repo/migrations/20160814154456_create_transaction.exs]: timestamps([{:inserted_at, :created_at}])

File: web/models/transaction_log.ex [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+    field :receiver_name, :string
+    field :receiver_gateway_name, :string
+    field :receiver_branch_name, :string
+    field :receiver_account_number, :string
+    field :request_data, :string
+    field :response_data, :string
+    field :transaction_id, :string
+    belongs_to :order, Delivr.Order
+
+    timestamps()
```
Comment (hieuphq) [inline web/models/transaction_log.ex]: timestamps([{:inserted_at, :created_at}])

File: web/controllers/order_controller.ex [hunk 7]
```diff
@@ -357,4 +378,86 @@ defmodule Delivr.OrderController do
...
```
Comment (hieuphq) [inline web/controllers/order_controller.ex]: remove 2 empty lines

File: web/controllers/order_controller.ex [hunk 8]
```diff
@@ -357,4 +378,86 @@ defmodule Delivr.OrderController do
     end
   end
 
+  def generate_transaction_log_for_stripe(order_id, request_data, response_data) do
+    order = Repo.get!(Order, order_id)
+      |> Repo.preload([:order_items, :voucher])
+
+    if order do
+      {_status, request_json} = JSON.encode(request_data)
+
```
Comment (hieuphq) [inline web/controllers/order_controller.ex]: remove empty line

**Comment by hieuphq** [comment]
@thminhVN 👍 can merge to develop

**Pull request merged**